# Check Homework HW04

In [185]:
import pandas as pd
import numpy as np

Now, import your solutions from `hw4_answers.py`. The following code looks a bit redundant. However, we do this to allow reloading the `hw4_answers.py` in case you made some changes. Normally, Python assumes that modules don't change and therefore does not try to import them again.

In [ ]:
import hw4_answers
reload(hw4_answers)
from hw4_answers import *

## Problem 1
Create a function `load_employees` that loads the employees table from
the file  `/home/data/AdventureWorks/Employees.xls` and sets the index of the DataFrame to the `EmployeeID`. The function should return a table with the `EmployeeID` as the index and the remaining 25 columns.

In [190]:
# step1 load the data from path
employee_df = pd.read_excel('/home/data/AdventureWorks/Employees.xls')
employees_df.head()

EmployeeID  ManagerID  TerritoryID Title FirstName MiddleName LastName  \
0         259      250.0          NaN   NaN       Ben          T   Miller   
1         278      274.0          6.0   NaN   Garrett          R   Vargas   
2         204       26.0          NaN   NaN      Gabe          B    Mares   
3          78       26.0          NaN   NaN    Reuben          H     D'sa   
4         255      250.0          NaN   NaN    Gordon          L      Hee   

  Suffix                      JobTitle  NationalIDNumber  ... SickLeaveHours  \
0    NaN                         Buyer          20269531  ...             47   
1    NaN          Sales Representative         234474252  ...             36   
2    NaN  Production Technician - WC40         440379437  ...             48   
3    NaN  Production Supervisor - WC40         370989364  ...             56   
4    NaN                         Buyer         466142721  ...             46   

    PhoneNumber PhoneNumberType                 EmailAddress  \
0  151-555-0113            Work     ben0@adventure-works.com   
1  922-555-0165            Work    garrett1@mapleleafmail.ca   
2  310-555-0117            Work    gabe0@adventure-works.com   
3  191-555-0112            Work  reuben0@adventure-works.com   
4  230-555-0144            Cell  gordon0@adventure-works.com   

           AddressLine1  AddressLine2      City StateProvinceName PostalCode  \
0         101 Candy Rd.           NaN   Redmond        Washington      98052   
1    10203 Acorn Avenue           NaN   Calgary           Alberta    T2P 2G8   
2   1061 Buskrik Avenue           NaN   Edmonds        Washington      98020   
3  1064 Slow Creek Road           NaN   Seattle        Washington      98104   
4    108 Lakeside Court           NaN  Bellevue        Washington      98004   

     CountryName  
0  United States  
1         Canada  
2  United States  
3  United States  
4  United States  

[5 rows x 26 columns]

In [191]:
# step 2: change index 
employee_df = employee_df.set_index('EmployeeID')
# employee_df.columns 
employee_df.index

Int64Index([259, 278, 204,  78, 255,  66, 270,  22, 161, 124,
            ...
            123,  29, 101, 221,  18,  46, 202, 289, 288, 999],
           dtype='int64', name='EmployeeID', length=291)

In [192]:
def load_employees(file_path): 
    df = pd.read_excel(file_path)
    df = df.set_index('EmployeeID')
    return df


In [193]:
employees_df = load_employees('/home/data/AdventureWorks/Employees.xls')

print("Number of rows: %d\nNumber of cols: %d\n" % (employees_df.shape[0], employees_df.shape[1]))
print("Head of index: %s\n" % (employees_df.index[:10]))
print("Record of employee with ID=999\n")
print(employees_df.loc[999])

Number of rows: 291
Number of cols: 25

Head of index: Int64Index([259, 278, 204, 78, 255, 66, 270, 22, 161, 124], dtype='int64', name='EmployeeID')

Record of employee with ID=999

ManagerID                                      1
TerritoryID                                  NaN
Title                                        NaN
FirstName                               Chadwick
MiddleName                                   NaN
LastName                                   Smith
Suffix                                       NaN
JobTitle                            BI Professor
NationalIDNumber                       123456789
BirthDate                             1967-07-05
MaritalStatus                                  M
Gender                                         M
HireDate              2003-12-31 23:59:59.997000
SalariedFlag                                   0
VacationHours                                 55
SickLeaveHours                                47
PhoneNumber                       

The output should be 
<pre>
Number of rows: 291
Number of cols: 25

Head of index: Int64Index([259, 278, 204, 78, 255, 66, 270, 22, 161, 124], dtype='int64', name=u'EmployeeID')

Record of employee with ID=999

ManagerID                                      1
TerritoryID                                  NaN
Title                                        NaN
FirstName                               Chadwick
MiddleName                                   NaN
LastName                                   Smith
Suffix                                       NaN
JobTitle                            BI Professor
NationalIDNumber                       123456789
BirthDate                             1967-07-05
MaritalStatus                                  M
Gender                                         M
HireDate              2003-12-31 23:59:59.997000
SalariedFlag                                   0
VacationHours                                 55
SickLeaveHours                                47
PhoneNumber                         555-887-9788
PhoneNumberType                             Work
EmailAddress         chadwick.smith@rentpath.com
AddressLine1                   565 Peachtree Rd.
AddressLine2                                 NaN
City                                     Atlanta
StateProvinceName                        Georgia
PostalCode                                 30084
CountryName                        United States
Name: 999, dtype: object
</pre>

The output should be 
<pre>
Number of rows: 291
Number of cols: 25

Head of index: Int64Index([259, 278, 204, 78, 255, 66, 270, 22, 161, 124], dtype='int64', name=u'EmployeeID')

Record of employee with ID=999

ManagerID                                      1
TerritoryID                                  NaN
Title                                        NaN
FirstName                               Chadwick
MiddleName                                   NaN
LastName                                   Smith
Suffix                                       NaN
JobTitle                            BI Professor
NationalIDNumber                       123456789
BirthDate                             1967-07-05
MaritalStatus                                  M
Gender                                         M
HireDate              2003-12-31 23:59:59.997000
SalariedFlag                                   0
VacationHours                                 55
SickLeaveHours                                47
PhoneNumber                         555-887-9788
PhoneNumberType                             Work
EmailAddress         chadwick.smith@rentpath.com
AddressLine1                   565 Peachtree Rd.
AddressLine2                                 NaN
City                                     Atlanta
StateProvinceName                        Georgia
PostalCode                                 30084
CountryName                        United States
Name: 999, dtype: object
</pre>

In [98]:
employees_df.head()

ManagerID  TerritoryID Title FirstName MiddleName LastName Suffix  \
EmployeeID                                                                      
259             250.0          NaN   NaN       Ben          T   Miller    NaN   
278             274.0          6.0   NaN   Garrett          R   Vargas    NaN   
204              26.0          NaN   NaN      Gabe          B    Mares    NaN   
78               26.0          NaN   NaN    Reuben          H     D'sa    NaN   
255             250.0          NaN   NaN    Gordon          L      Hee    NaN   

                                JobTitle  NationalIDNumber   BirthDate  ...  \
EmployeeID                                                              ...   
259                                Buyer          20269531  1967-07-05  ...   
278                 Sales Representative         234474252  1969-03-07  ...   
204         Production Technician - WC40         440379437  1982-06-11  ...   
78          Production Supervisor - WC40         370989364  1981-09-27  ...   
255                                Buyer         466142721  1960-12-30  ...   

           SickLeaveHours   PhoneNumber PhoneNumberType  \
EmployeeID                                                
259                    47  151-555-0113            Work   
278                    36  922-555-0165            Work   
204                    48  310-555-0117            Work   
78                     56  191-555-0112            Work   
255                    46  230-555-0144            Cell   

                           EmailAddress          AddressLine1  AddressLine2  \
EmployeeID                                                                    
259            ben0@adventure-works.com         101 Candy Rd.           NaN   
278           garrett1@mapleleafmail.ca    10203 Acorn Avenue           NaN   
204           gabe0@adventure-works.com   1061 Buskrik Avenue           NaN   
78          reuben0@adventure-works.com  1064 Slow Creek Road           NaN   
255         gordon0@adventure-works.com    108 Lakeside Court           NaN   

                City StateProvinceName PostalCode    CountryName  
EmployeeID                                                        
259          Redmond        Washington      98052  United States  
278          Calgary           Alberta    T2P 2G8         Canada  
204          Edmonds        Washington      98020  United States  
78           Seattle        Washington      98104  United States  
255         Bellevue        Washington      98004  United States  

[5 rows x 25 columns]

### Slicing
* picking out *part of the DF* (rows, columns, elements)

#### Pick out rows
* pick by row names -- LOC
* pick by row numbers -- ILOC: I stands for integer

#### Pick out columns
* pick by column names -- LOC (COMMA)
* pick by column numbers -- ILOC: I stands for integer

In [85]:
# picking rows by rowname -- LOC 
my_columns = ['FirstName', 'MiddleName', 'LastName']
employee_df.loc[259, my_columns]

IndexingError: Too many indexers

In [87]:
## ILOC: I means INTEGER 
my_columns = ['FirstName', 'MiddleName', 'LastName']
employee_df.iloc[259, :]  ## row 260th 

ManagerID                                      26
TerritoryID                                   NaN
Title                                         NaN
FirstName                                  Sidney
MiddleName                                      M
LastName                                     Higa
Suffix                                        NaN
JobTitle             Production Technician - WC10
NationalIDNumber                        844973625
BirthDate                              1950-10-01
MaritalStatus                                   M
Gender                                          M
HireDate                      2002-03-05 00:00:00
SalariedFlag                                    0
VacationHours                                  84
SickLeaveHours                                 62
PhoneNumber                          424-555-0189
PhoneNumberType                              Work
EmailAddress          sidney0@adventure-works.com
AddressLine1               9277 Country View Lane


In [90]:
employee_df.iat[259, 10]   # element on row 260 and column 11

'M'

In [35]:
employees_df.iloc[2, 1:4]

TerritoryID     NaN
Title           NaN
FirstName      Gabe
Name: 204, dtype: object

In [26]:
employees_df['FullName'] = employees_df['LastName'].astype(str) + ", " + employees_df['FirstName'].astype(str) + " " + employees_df['MiddleName'].astype(str)
employees_df.head()

ManagerID  TerritoryID Title FirstName MiddleName LastName Suffix  \
EmployeeID                                                                      
259             250.0          NaN   NaN       Ben          T   Miller    NaN   
278             274.0          6.0   NaN   Garrett          R   Vargas    NaN   
204              26.0          NaN   NaN      Gabe          B    Mares    NaN   
78               26.0          NaN   NaN    Reuben          H     D'sa    NaN   
255             250.0          NaN   NaN    Gordon          L      Hee    NaN   

                                JobTitle  NationalIDNumber   BirthDate  ...  \
EmployeeID                                                              ...   
259                                Buyer          20269531  1967-07-05  ...   
278                 Sales Representative         234474252  1969-03-07  ...   
204         Production Technician - WC40         440379437  1982-06-11  ...   
78          Production Supervisor - WC40         370989364  1981-09-27  ...   
255                                Buyer         466142721  1960-12-30  ...   

             PhoneNumber PhoneNumberType                 EmailAddress  \
EmployeeID                                                              
259         151-555-0113            Work     ben0@adventure-works.com   
278         922-555-0165            Work    garrett1@mapleleafmail.ca   
204         310-555-0117            Work    gabe0@adventure-works.com   
78          191-555-0112            Work  reuben0@adventure-works.com   
255         230-555-0144            Cell  gordon0@adventure-works.com   

                    AddressLine1  AddressLine2      City StateProvinceName  \
EmployeeID                                                                   
259                101 Candy Rd.           NaN   Redmond        Washington   
278           10203 Acorn Avenue           NaN   Calgary           Alberta   
204          1061 Buskrik Avenue           NaN   Edmonds        Washington   
78          1064 Slow Creek Road           NaN   Seattle        Washington   
255           108 Lakeside Court           NaN  Bellevue        Washington   

           PostalCode    CountryName           FullName  
EmployeeID                                               
259             98052  United States      Miller, Ben T  
278           T2P 2G8         Canada  Vargas, Garrett R  
204             98020  United States      Mares, Gabe B  
78              98104  United States     D'sa, Reuben H  
255             98004  United States      Hee, Gordon L  

[5 rows x 26 columns]

## Problem 2
Define a function `getFullName` which takes the employees table and a single employee ID as arguments, and returns a string with the full name of the employee in the format "LAST, FIRST MIDDLE".
If the given ID does not belong to any employee return the string "UNKNOWN" (in all caps)
If no middle name is given only return "LAST, FIRST". Make sure there are not trailing spaces!
If only the middle initial is given the return the full name in the format "LAST, FIRST M." with the middle initial followed by a '.'.

Arguments:
- df (DataFrame): Employee Table
- empid (int):  Employee ID

Returns:
- String with full name

In [194]:
employee_df.columns

Index(['ManagerID', 'TerritoryID', 'Title', 'FirstName', 'MiddleName',
       'LastName', 'Suffix', 'JobTitle', 'NationalIDNumber', 'BirthDate',
       'MaritalStatus', 'Gender', 'HireDate', 'SalariedFlag', 'VacationHours',
       'SickLeaveHours', 'PhoneNumber', 'PhoneNumberType', 'EmailAddress',
       'AddressLine1', 'AddressLine2', 'City', 'StateProvinceName',
       'PostalCode', 'CountryName'],
      dtype='object')

In [195]:
# step 1: create new column cuz start simple 
employee_df['FullName'] = employee_df['LastName'] + ', ' + employee_df['FirstName'] + ' ' + employee_df['MiddleName']

In [207]:
# step 2: given a emp_id, return full_name
empid = 10000
employee_df.loc[empid, 'FullName']



KeyError: 10000

In [211]:
mydf= pd.read_excel('/home/data/AdventureWorks/Employees.xls')
mydf.columns

Index(['EmployeeID', 'ManagerID', 'TerritoryID', 'Title', 'FirstName',
       'MiddleName', 'LastName', 'Suffix', 'JobTitle', 'NationalIDNumber',
       'BirthDate', 'MaritalStatus', 'Gender', 'HireDate', 'SalariedFlag',
       'VacationHours', 'SickLeaveHours', 'PhoneNumber', 'PhoneNumberType',
       'EmailAddress', 'AddressLine1', 'AddressLine2', 'City',
       'StateProvinceName', 'PostalCode', 'CountryName'],
      dtype='object')

In [212]:
mydf.head()

EmployeeID  ManagerID  TerritoryID Title FirstName MiddleName LastName  \
0         259      250.0          NaN   NaN       Ben          T   Miller   
1         278      274.0          6.0   NaN   Garrett          R   Vargas   
2         204       26.0          NaN   NaN      Gabe          B    Mares   
3          78       26.0          NaN   NaN    Reuben          H     D'sa   
4         255      250.0          NaN   NaN    Gordon          L      Hee   

  Suffix                      JobTitle  NationalIDNumber  ... SickLeaveHours  \
0    NaN                         Buyer          20269531  ...             47   
1    NaN          Sales Representative         234474252  ...             36   
2    NaN  Production Technician - WC40         440379437  ...             48   
3    NaN  Production Supervisor - WC40         370989364  ...             56   
4    NaN                         Buyer         466142721  ...             46   

    PhoneNumber PhoneNumberType                 EmailAddress  \
0  151-555-0113            Work     ben0@adventure-works.com   
1  922-555-0165            Work    garrett1@mapleleafmail.ca   
2  310-555-0117            Work    gabe0@adventure-works.com   
3  191-555-0112            Work  reuben0@adventure-works.com   
4  230-555-0144            Cell  gordon0@adventure-works.com   

           AddressLine1  AddressLine2      City StateProvinceName PostalCode  \
0         101 Candy Rd.           NaN   Redmond        Washington      98052   
1    10203 Acorn Avenue           NaN   Calgary           Alberta    T2P 2G8   
2   1061 Buskrik Avenue           NaN   Edmonds        Washington      98020   
3  1064 Slow Creek Road           NaN   Seattle        Washington      98104   
4    108 Lakeside Court           NaN  Bellevue        Washington      98004   

     CountryName  
0  United States  
1         Canada  
2  United States  
3  United States  
4  United States  

[5 rows x 26 columns]

In [210]:
# step 3: if else to handle empid not in df
empid = 259
empid in employee_df.index

True

In [225]:
# step 4: pick out f, m, l
employees_df = load_employees('/home/data/AdventureWorks/Employees.xls')

empid = 274

def getFullName(df, empid):
    if empid in df.index:
        ln = df.loc[empid, 'LastName'] 
        fn = df.loc[empid, 'FirstName'] 
        mn = df.loc[empid, 'MiddleName']
        my_return = ln + ', ' + fn + ' ' + mn
        return my_return
    else:
        return 'UNKNOWN'


In [217]:
#step 5: wrap up as an func:
def getFullName(df, empid):
    if empid in df.FullName:
        full_name = df.loc[empid, ['LastName']] + ', ' + df.loc[empid, ['FirstName']] + ' ' + df.loc[empid, ['MiddleName']
        my_return = full_name.loc[empid, 'FullName']
        return my_return
    else:
        return 'UNKNOWN'

In [231]:
eid = 4
getFullName(employee_df, eid)
# employee_df


TypeError: must be str, not float

In [232]:
employee_df.loc[4, :]

ManagerID                              25
TerritoryID                           NaN
Title                                 NaN
FirstName                             Rob
MiddleName                            NaN
LastName                          Walters
Suffix                                NaN
JobTitle             Senior Tool Designer
NationalIDNumber                112457891
BirthDate                      1969-01-23
MaritalStatus                           S
Gender                                  M
HireDate              2002-01-05 00:00:00
SalariedFlag                            0
VacationHours                          48
SickLeaveHours                         80
PhoneNumber                  612-555-0100
PhoneNumberType                      Cell
EmailAddress                 rob0@aol.com
AddressLine1          5678 Lakeview Blvd.
AddressLine2                          NaN
City                          Minneapolis
StateProvinceName               Minnesota
PostalCode                        

In [197]:
employee_df.loc[259, 'FullName']

'Miller, Ben T'

In [199]:
employee_df.iloc[259, 10]  ## row 260 column 11 b


'M'

In [201]:
###  loc: column or row
###  at: element 
employee_df.iat[259, 10]  # i: integer -- row & column numbers 

employee_df.at[259, 'FullName']  # row & column names 

'Miller, Ben T'

In [205]:
employee_df1 = employee_df.copy().reset_index()
employee_df1.iloc[0, :]

EmployeeID                                259
ManagerID                                 250
TerritoryID                               NaN
Title                                     NaN
FirstName                                 Ben
MiddleName                                  T
LastName                               Miller
Suffix                                    NaN
JobTitle                                Buyer
NationalIDNumber                     20269531
BirthDate                          1967-07-05
MaritalStatus                               M
Gender                                      M
HireDate                  2004-04-09 00:00:00
SalariedFlag                                0
VacationHours                              55
SickLeaveHours                             47
PhoneNumber                      151-555-0113
PhoneNumberType                          Work
EmailAddress         ben0@adventure-works.com
AddressLine1                    101 Candy Rd.
AddressLine2                      

In [108]:
# step1: pick out f, m, l name based on e_id

empid = 100000
# my_columns = ["FirstName","MiddleName","LastName"]
# name = employee_df.loc[empid, my_columns]  ## pandas series 
# name

# # step2: assembly full name in the format as requested from f, m, l in step 1
# fn = name[0]
# mn = name[1]
# ln = name[2]
# full_name = ln + ", " + fn + " " + mn
# full_name

def getFullName(df, empid):
    # handle UNKNOWN
    if empid in df.index:
        my_columns = ["FirstName","MiddleName","LastName"]
        name = employee_df.loc[empid, my_columns]  ## pandas series 
        name

        # step2: assembly full name in the format as requested from f, m, l in step 1
        fn = name[0]
        mn = name[1]
        ln = name[2]
        full_name = ln + ", " + fn + " " + mn
        return full_name
    else:
        return "UNKNOWN"

In [38]:
def getFullName(df, empid):
    full_name = "UNKNOWN"
    if empid in employees_df.index:
        name = df.loc[empid, ["FirstName","MiddleName","LastName"]]
        first_name, middle_name, last_name = name[0],name[1],name[2]
        if str(middle_name) == 'nan':
            full_name = last_name +", " + first_name
        elif len(middle_name) == 1:
            full_name = last_name + ", " + first_name + " " + middle_name + "."
        else:
            full_name = last_name + ", " + first_name + " " + middle_name

    return full_name

In [214]:
for eid in [274]:
    print('%d, "%s"' %(eid, getFullName(employees_df, eid)))

AttributeError: 'DataFrame' object has no attribute 'FullName'

The output should be
<pre>
274, "Jiang, Stephen Y."
999, "Smith, Chadwick"
102, "Mu, Zheng W."
</pre>

## Problem 3
Define a function `isSales` that takes the job title of an employee as string as an argument and return either `True` if the job title indicates this person works in sales, and `False` otherwise.

Argument:
- jobtitle (str)

Returns:
- True or False

In [233]:
def isSales(jobtitle):
    if "sale" in jobtitle.lower(): ## SALE sALe
        return True
    else:
        return False

In [234]:
for jt in ['Chief Data Scientist', 'Sales Manager', 'Vice President of Sales']:
    if isSales(jt):
        print("The job title '%s' is part of the Sales Department." % jt)
    else:
        print("The job title '%s' belongs to a different department." % jt)

The job title 'Chief Data Scientist' belongs to a different department.
The job title 'Sales Manager' is part of the Sales Department.
The job title 'Vice President of Sales' is part of the Sales Department.


The output should be
<pre>
The job title 'Chief Data Scientist' belongs to a different department.
The job title 'Sales Manager' is part of the Sales Department.
The job title 'Vice President of Sales' is part of the Sales Department.
</pre>

In [123]:
employee_df['JobTitle']

EmployeeID
259                                  Buyer
278                   Sales Representative
204           Production Technician - WC40
78            Production Supervisor - WC40
255                                  Buyer
66            Production Technician - WC60
270                 Database Administrator
22                    Marketing Specialist
161           Production Technician - WC45
124                                Stocker
10        Research and Development Manager
248                             Accountant
175           Production Technician - WC30
155           Production Technician - WC20
144           Production Technician - WC20
192           Production Supervisor - WC40
166           Production Supervisor - WC30
112           Production Technician - WC50
203           Production Technician - WC40
245                             Accountant
257                                  Buyer
244         Accounts Receivable Specialist
69            Production Technician - WC60


## Problem 4
Define a function `filterSales` with the employee tables as an argument, that returns a new table of the same column_names and row_names containing only row of sales people. You should use the `isSales` function from the previous problem.

Arguments:
- employees (DataFrame)

Returns:
- DataFrame with only people form the Sales Department

In [246]:
my_column_list = ['FirstName', 'LastName']
employee_df.loc[:, my_column_list]

FirstName           LastName
EmployeeID                              
259               Ben             Miller
278           Garrett             Vargas
204              Gabe              Mares
78             Reuben               D'sa
255            Gordon                Hee
66              Karan             Khanna
270          François           Ajenstat
22             Sariya  Harnpadoungsataya
161              Kirk       Koenigsbauer
124               Kim              Ralls
10            Michael             Raheem
248              Mike            Seamans
175              Reed               Koch
155              Fadi           Fakhouri
144              Paul              Singh
192            Brenda               Diaz
166              Jack            Richins
112              John              Evans
203               Ken               Myer
245           Barbara           Moreland
257              Eric             Kurjan
244             Bryan             Walton
69              Steve            Masters
210           Belinda             Newman
163              Alex            Nayberg
118               Don               Hall
176             David           Lawrence
147            Sandra     Reátegui Alayo
138          Samantha              Smith
95                Jim          Scardelis
...               ...                ...
70              David              Ortiz
219              Sean               Chai
148             Jason            Watters
140          Prasanna     Samarawickrama
290            Ranjit  Varkey Chudukatil
5                Gail           Erickson
252            Arvind                Rao
266             Peter           Connelly
233            Magnus            Hedlund
83            Patrick               Cook
251            Mikael           Sandberg
241             David                Liu
25              James           Hamilton
108           Jinghao                Liu
239             Mindy             Martin
27                 Jo              Brown
186             Shane                Kim
38                Kim        Abercrombie
154           Raymond                Sam
114         Mindaugas         Krapauskas
123             Vamsi              Kuppa
29               Mark           McArthur
101            Houman         Pournasseh
221             Chris             Norred
18               John               Wood
46             Eugene              Kogan
202            Tawana            Nusbaum
289               Jae                Pak
288            Rachel             Valdez
999          Chadwick              Smith

[291 rows x 2 columns]

In [239]:
# step 1: picking out rows by jobtitle 
# employee_df.columns

sales_filter = employee_df.loc[:, ['JobTitle']].str.lower().str.contains('sale')
# my_con

EmployeeID
259    False
278     True
204    False
78     False
255    False
66     False
270    False
22     False
161    False
124    False
10     False
248    False
175    False
155    False
144    False
192    False
166    False
112    False
203    False
245    False
257    False
244    False
69     False
210    False
163    False
118    False
176    False
147    False
138    False
95     False
       ...  
70     False
219    False
148    False
140    False
290     True
5      False
252    False
266    False
233    False
83     False
251    False
241    False
25     False
108    False
239    False
27     False
186    False
38     False
154    False
114    False
123    False
29     False
101    False
221    False
18     False
46     False
202    False
289     True
288     True
999    False
Name: JobTitle, Length: 291, dtype: bool

In [240]:
# step 2: pick out rows based on the filter then assign to a new DF
salesDF = employee_df[sales_filter]
salesDF

ManagerID  TerritoryID Title FirstName MiddleName  \
EmployeeID                                                      
278             274.0          6.0   NaN   Garrett          R   
283             274.0          1.0   NaN     David          R   
274             273.0          NaN   NaN   Stephen          Y   
276             274.0          4.0   NaN     Linda          C   
286             285.0          9.0   NaN      Lynn          N   
284             274.0          1.0   Mr.      Tete          A   
287             273.0          NaN   NaN       Amy          E   
281             274.0          4.0   NaN       Shu          K   
280             274.0          1.0   NaN    Pamela          O   
285             273.0          NaN   Mr.      Syed          E   
273               1.0          NaN   Mr.     Brian          S   
277             274.0          3.0   NaN   Jillian        NaN   
275             274.0          2.0   NaN   Michael          G   
279             274.0          5.0   NaN      Tsvi    Michael   
282             274.0          6.0   NaN      José    Edvaldo   
290             287.0          7.0   NaN    Ranjit          R   
289             287.0         10.0   NaN       Jae          B   
288             287.0          8.0   NaN    Rachel          B   

                     LastName Suffix                      JobTitle  \
EmployeeID                                                           
278                    Vargas    NaN          Sales Representative   
283                  Campbell    NaN          Sales Representative   
274                     Jiang    NaN  North American Sales Manager   
276                  Mitchell    NaN          Sales Representative   
286                  Tsoflias    NaN          Sales Representative   
284               Mensa-Annan    NaN          Sales Representative   
287                   Alberts    NaN        European Sales Manager   
281                       Ito    NaN          Sales Representative   
280              Ansman-Wolfe    NaN          Sales Representative   
285                     Abbas    NaN         Pacific Sales Manager   
273                   Welcker    NaN       Vice President of Sales   
277                    Carson    NaN          Sales Representative   
275                    Blythe    NaN          Sales Representative   
279                    Reiter    NaN          Sales Representative   
282                   Saraiva    NaN          Sales Representative   
290         Varkey Chudukatil    NaN          Sales Representative   
289                       Pak    NaN          Sales Representative   
288                    Valdez    NaN          Sales Representative   

            NationalIDNumber   BirthDate  ...          PhoneNumber  \
EmployeeID                                ...                        
278                234474252  1969-03-07  ...         922-555-0165   
283                987554265  1968-03-14  ...         740-555-0182   
274                502097814  1945-11-17  ...         238-555-0197   
276                191644724  1974-03-30  ...         883-555-0116   
286                758596752  1965-04-18  ...  1 (11) 500 555-0190   
284                 90836195  1972-02-06  ...         615-555-0153   
287                982310417  1951-10-22  ...         775-555-0164   
281                139397894  1962-04-10  ...         330-555-0120   
280                 61161660  1969-01-06  ...         340-555-0193   
285                481044938  1969-02-11  ...         926-555-0182   
273                112432117  1971-07-08  ...         716-555-0127   
277                615389812  1956-09-29  ...         517-555-0117   
275                841560125  1963-01-26  ...         257-555-0154   
279                716374314  1968-02-19  ...         664-555-0112   
282                399771412  1958-01-11  ...         185-555-0169   
290                134219713  1969-10-31  ...  1 (11) 500 555-0117   
289                668991357  1962-04-18  ...  1 (

### for loops in DataFrame
* step 1 loop through the rows
    * pick out job title, then call function isSales
    * if true:
        * store in the new DF


### make it better
* apply -- for loop on DF & PD.Series
* map -- for loop on PD.Series
* applymap -- for loop on elements of DF & PD.Series  -- much less used because it is VERY slow
    

In [126]:
sales_filter = employees_df.loc[:, "JobTitle"].apply(isSales)
sales_filter

EmployeeID
259    False
278     True
204    False
78     False
255    False
66     False
270    False
22     False
161    False
124    False
10     False
248    False
175    False
155    False
144    False
192    False
166    False
112    False
203    False
245    False
257    False
244    False
69     False
210    False
163    False
118    False
176    False
147    False
138    False
95     False
       ...  
70     False
219    False
148    False
140    False
290     True
5      False
252    False
266    False
233    False
83     False
251    False
241    False
25     False
108    False
239    False
27     False
186    False
38     False
154    False
114    False
123    False
29     False
101    False
221    False
18     False
46     False
202    False
289     True
288     True
999    False
Name: JobTitle, Length: 291, dtype: bool

In [241]:
def filterSales(employees):
    sales_filter = employees.loc[:, "JobTitle"].apply(isSales)
    return employees[sales_filter]

In [242]:
sales_df = filterSales(employees_df)
print("Number of rows: %d\nNumber of cols: %d\n" % (sales_df.shape[0], sales_df.shape[1]))
print("Head of index: %s\n" % (sales_df.index[:10]))
print("Record of sales employee with ID=280\n")
print(sales_df.loc[280])

Number of rows: 18
Number of cols: 25

Head of index: Int64Index([278, 283, 274, 276, 286, 284, 287, 281, 280, 285], dtype='int64', name='EmployeeID')

Record of sales employee with ID=280

ManagerID                             274
TerritoryID                             1
Title                                 NaN
FirstName                          Pamela
MiddleName                              O
LastName                     Ansman-Wolfe
Suffix                                NaN
JobTitle             Sales Representative
NationalIDNumber                 61161660
BirthDate                      1969-01-06
MaritalStatus                           S
Gender                                  F
HireDate              2005-10-01 00:00:00
SalariedFlag                            1
VacationHours                          22
SickLeaveHours                         31
PhoneNumber                  340-555-0193
PhoneNumberType                      Cell
EmailAddress            pamela0@yahoo.com
AddressLine1

The output should be
<pre>
Number of rows: 18
Number of cols: 25

Head of index: Int64Index([278, 283, 274, 276, 286, 284, 287, 281, 280, 285], dtype='int64', name=u'EmployeeID')

Record of sales employee with ID=280

ManagerID                             274
TerritoryID                             1
Title                                 NaN
FirstName                          Pamela
MiddleName                              O
LastName                     Ansman-Wolfe
Suffix                                NaN
JobTitle             Sales Representative
NationalIDNumber                 61161660
BirthDate                      1969-01-06
MaritalStatus                           S
Gender                                  F
HireDate              2005-10-01 00:00:00
SalariedFlag                            1
VacationHours                          22
SickLeaveHours                         31
PhoneNumber                  340-555-0193
PhoneNumberType                      Cell
EmailAddress            pamela0@yahoo.com
AddressLine1            636 Vine Hill Way
AddressLine2                          NaN
City                             Portland
StateProvinceName                  Oregon
PostalCode                          97205
CountryName                 United States
Name: 280, dtype: object
</pre>


## Problem 5
Define a function `getEmailList` with that returns a Series of strings of all email addresses of employees in this state or province. The email addresses should be separated by a given character, usually a comma ',' or semicolon ';'.

Arguments:
- employees (DataFrame)
- delimiter (str)

Returns:
- Series of email addresses, concatenated by the given delimiter. The Series is indexed by the state or province.

In [249]:
employee_df.columns

Index(['ManagerID', 'TerritoryID', 'Title', 'FirstName', 'MiddleName',
       'LastName', 'Suffix', 'JobTitle', 'NationalIDNumber', 'BirthDate',
       'MaritalStatus', 'Gender', 'HireDate', 'SalariedFlag', 'VacationHours',
       'SickLeaveHours', 'PhoneNumber', 'PhoneNumberType', 'EmailAddress',
       'AddressLine1', 'AddressLine2', 'City', 'StateProvinceName',
       'PostalCode', 'CountryName', 'FullName'],
      dtype='object')

In [251]:
employee_df.head()

ManagerID  TerritoryID Title FirstName MiddleName LastName Suffix  \
EmployeeID                                                                      
259             250.0          NaN   NaN       Ben          T   Miller    NaN   
278             274.0          6.0   NaN   Garrett          R   Vargas    NaN   
204              26.0          NaN   NaN      Gabe          B    Mares    NaN   
78               26.0          NaN   NaN    Reuben          H     D'sa    NaN   
255             250.0          NaN   NaN    Gordon          L      Hee    NaN   

                                JobTitle  NationalIDNumber   BirthDate  ...  \
EmployeeID                                                              ...   
259                                Buyer          20269531  1967-07-05  ...   
278                 Sales Representative         234474252  1969-03-07  ...   
204         Production Technician - WC40         440379437  1982-06-11  ...   
78          Production Supervisor - WC40         370989364  1981-09-27  ...   
255                                Buyer         466142721  1960-12-30  ...   

             PhoneNumber PhoneNumberType                 EmailAddress  \
EmployeeID                                                              
259         151-555-0113            Work     ben0@adventure-works.com   
278         922-555-0165            Work    garrett1@mapleleafmail.ca   
204         310-555-0117            Work    gabe0@adventure-works.com   
78          191-555-0112            Work  reuben0@adventure-works.com   
255         230-555-0144            Cell  gordon0@adventure-works.com   

                    AddressLine1  AddressLine2      City StateProvinceName  \
EmployeeID                                                                   
259                101 Candy Rd.           NaN   Redmond        Washington   
278           10203 Acorn Avenue           NaN   Calgary           Alberta   
204          1061 Buskrik Avenue           NaN   Edmonds        Washington   
78          1064 Slow Creek Road           NaN   Seattle        Washington   
255           108 Lakeside Court           NaN  Bellevue        Washington   

           PostalCode    CountryName           FullName  
EmployeeID                                               
259             98052  United States      Miller, Ben T  
278           T2P 2G8         Canada  Vargas, Garrett R  
204             98020  United States      Mares, Gabe B  
78              98104  United States     D'sa, Reuben H  
255             98004  United States      Hee, Gordon L  

[5 rows x 26 columns]

In [250]:
state = 'Washington'
my_filter = employee_df['StateProvinceName'] == state
my_filter

EmployeeID
259     True
278    False
204     True
78      True
255     True
66      True
270     True
22      True
161     True
124     True
10     False
248     True
175     True
155     True
144     True
192     True
166     True
112     True
203    False
245     True
257     True
244    False
69      True
210     True
163     True
118     True
176     True
147     True
138     True
95      True
       ...  
70      True
219     True
148     True
140     True
290    False
5       True
252     True
266     True
233     True
83      True
251     True
241     True
25      True
108     True
239     True
27      True
186     True
38      True
154     True
114     True
123     True
29      True
101     True
221     True
18      True
46      True
202     True
289    False
288    False
999    False
Name: StateProvinceName, Length: 291, dtype: bool

In [252]:
# step 1: get emails based on state
# employee_df['StateProvinceName']
state = 'Washington'
my_filter = employee_df['StateProvinceName'] == state
my_columns =  ['EmailAddress', 'StateProvinceName']
employee_df.loc[my_filter, my_columns]

EmailAddress StateProvinceName
EmployeeID                                                
259             ben0@adventure-works.com        Washington
204            gabe0@adventure-works.com        Washington
78           reuben0@adventure-works.com        Washington
255          gordon0@adventure-works.com        Washington
66                    karan0@hotmail.com        Washington
270                  françois0@yahoo.com        Washington
22           sariya0@adventure-works.com        Washington
161                    kirk0@hotmail.com        Washington
124             kim0@adventure-works.com        Washington
248                  mike0@safe-mail.net        Washington
175                    reed0@hotmail.com        Washington
155            fadi0@adventure-works.com        Washington
144                  paul1@safe-mail.net        Washington
192          brenda0@adventure-works.com        Washington
166            jack0@adventure-works.com        Washington
112                      john1@gmail.com        Washington
245         barbara1@adventure-works.com        Washington
257                    eric2@hotmail.com        Washington
69            steve0@adventure-works.com        Washington
210         belinda0@adventure-works.com        Washington
163            alex0@adventure-works.com        Washington
118             don0@adventure-works.com        Washington
176           david7@adventure-works.com        Washington
147          sandra0@adventure-works.com        Washington
138              samantha0@safe-mail.net        Washington
95                    jim0@safe-mail.net        Washington
231              jo1@adventure-works.com        Washington
86             ryan0@adventure-works.com        Washington
214         andreas0@adventure-works.com        Washington
72                 steven0@safe-mail.net        Washington
...                                  ...               ...
169           susan1@adventure-works.com        Washington
209          kathie0@adventure-works.com        Washington
50           sidney0@adventure-works.com        Washington
222                      ascott0@aol.com        Washington
70            david2@adventure-works.com        Washington
219            sean1@adventure-works.com        Washington
148                       jason0@aol.com        Washington
140              prasanna0@safe-mail.net        Washington
5                        gail0@gmail.com        Washington
252          arvind0@adventure-works.com        Washington
266           peter1@adventure-works.com        Washington
233          magnus0@adventure-works.com        Washington
83          patrick1@adventure-works.com        Washington
251                mikael0@safe-mail.net        Washington
241           david6@adventure-works.com        Washington
25            james1@adventure-works.com        Washington
108         jinghao0@adventure-works.com        Washington
239           mindy0@adventure-works.com        Washington
27               jo0@adventure-works.com        Washington
186                   shane0@hotmail.com        Washington
38                        kim1@yahoo.com        Washington
154               raymond0@safe-mail.net        Washington
114               mindaugas0@hotmail.com        Washington
123                   vamsi0@hotmail.com        Washington
29             mark1@adventure-works.com        Washington
101                      houman0@aol.com        Washington
221           chris1@adventure-works.com        Washington
18                         john5@aol.com        Washington
46                   eugene1@hotmail.com        Washington
202          tawana0@adventure-works.com        Washington

[271 rows x 2 columns]

In [258]:
my_email_list = employee_df.loc[my_filter, my_columns]
my_email_list = list(my_email_list['EmailAddress'])

In [260]:
delimiter = ', '
state + ": " + delimiter.join(my_email_list)

'Washington: ben0@adventure-works.com, gabe0@adventure-works.com, reuben0@adventure-works.com, gordon0@adventure-works.com, karan0@hotmail.com, françois0@yahoo.com, sariya0@adventure-works.com, kirk0@hotmail.com, kim0@adventure-works.com, mike0@safe-mail.net, reed0@hotmail.com, fadi0@adventure-works.com, paul1@safe-mail.net, brenda0@adventure-works.com, jack0@adventure-works.com, john1@gmail.com, barbara1@adventure-works.com, eric2@hotmail.com, steve0@adventure-works.com, belinda0@adventure-works.com, alex0@adventure-works.com, don0@adventure-works.com, david7@adventure-works.com, sandra0@adventure-works.com, samantha0@safe-mail.net, jim0@safe-mail.net, jo1@adventure-works.com, ryan0@adventure-works.com, andreas0@adventure-works.com, steven0@safe-mail.net, laura0@safe-mail.net, christopher0@adventure-works.com, dan0@adventure-works.com, yvonne0@adventure-works.com, michiko0@adventure-works.com, thierry0@adventure-works.com, michael2@aol.com, britta0@safe-mail.net, lorraine0@adventure-w

In [262]:
def getEmailListByState(employees, delimiter):
#     state = 'Washington'
    states = employee_df['StateProvinceName']
    output = dict()
    for state in states:
        my_filter = employee_df['StateProvinceName'] == state
        my_columns =  ['EmailAddress', 'StateProvinceName']
        my_email_df = employee_df.loc[my_filter, my_columns]
        my_email_list = list(my_email_df['EmailAddress'])
        my_output = delimiter.join(my_email_list)
        output[state] = my_output
    return pd.DataFrame(output)

In [143]:
# step 1: get emails based on state
# employee_df['StateProvinceName']
state = 'Washington'
my_filter = employee_df['StateProvinceName'] == state
my_columns =  ['EmailAddress', 'StateProvinceName']
employee_df.loc[my_filter, my_columns]

# step 2: make the emails a list
my_emails = employee_df.loc[my_filter, ['EmailAddress', 'StateProvinceName']]
my_emails

my_email_lists = list(my_emails['EmailAddress'])

delimiter = ', '
delimiter.join(my_email_lists)

def getEmailListByState(df, delimiter):
    states = employee_df['StateProvinceName']
    output = dict()
    for state in states:
        my_filter = employee_df['StateProvinceName'] == state
        my_emails = employee_df.loc[my_filter, ['EmailAddress', 'StateProvinceName']]

        my_email_lists = list(my_emails['EmailAddress'])

        output[state] = delimiter.join(my_email_lists)
    return pd.DataFrame(output)
    

In [268]:
# step 1: pick out emails and state names
employee_emails = employees_df.loc[:, ["StateProvinceName", "EmailAddress"]]
employee_emails.head()

# step 2: group emails by state names
my_groups = employee_emails.groupby('StateProvinceName')
pd.DataFrame(my_groups['EmailAddress'])

# step 3: turn email group into a list
my_lists = my_groups['EmailAddress'].apply(list)

# step 4: concat the list in step 3 by the delimiter 
# my_output = my_lists['EmailAddress'].apply(lambda email_list: delimiter.join(email_list))

def getEmailListByState(df, delimiter):
    employee_emails = employees_df.loc[:, ["StateProvinceName", "EmailAddress"]]
    my_groups = employee_emails.groupby('StateProvinceName')
#     my_lists = my_groups['EmailAddress'].apply(list)
    my_output = my_groups['EmailAddress'].apply(lambda email_list: delimiter.join(email_list))
    
    return my_output


In [57]:
def getEmailListByState(employees, delimiter):
    employee_emails = employees.loc[:, ["StateProvinceName", "EmailAddress"]]
#     delimiter = delimiter
    state_emails = employee_emails.groupby("StateProvinceName")
    
    ["EmailAddress"].apply(lambda x: delimiter.join(x))
    return state_emails


In [271]:
emails= getEmailListByState(salesDF, ", ")
for state in sorted(emails.index):
#     print('hi')
    print("%15s: %s" % (state, emails[state]))

        Alberta: garrett1@mapleleafmail.ca, ken1@adventure-works.com, bryan1@aol.com, michael0@adventure-works.com, ivo0@safe-mail.net
     California: michael6@adventure-works.com, shu0@adventure-works.com
        England: jae0@aol.co.uk
        Georgia: chadwick.smith@rentpath.com
        Gironde: ranjit0@adventure-works.com
        Hamburg: rachel0@adventure-works.com
  Massachusetts: tete0@adventure-works.com
       Michigan: michael9@adventure-works.com
      Minnesota: rob0@aol.com, jillian0@adventure-works.com
        Ontario: josé1@safe-mail.net
         Oregon: pamela0@yahoo.com
      Tennessee: tsvi0@adventure-works.com
           Utah: linda3@adventure-works.com
       Victoria: lynn0@adventure-works.com
     Washington: ben0@adventure-works.com, gabe0@adventure-works.com, reuben0@adventure-works.com, gordon0@adventure-works.com, karan0@hotmail.com, françois0@yahoo.com, sariya0@adventure-works.com, kirk0@hotmail.com, kim0@adventure-works.com, mike0@safe-mail.net, reed0@hotma

The output should be
<pre>
        Alberta: garrett1@mapleleafmail.ca
     California: shu0@adventure-works.com
        England: jae0@aol.co.uk
        Gironde: ranjit0@adventure-works.com
        Hamburg: rachel0@adventure-works.com
  Massachusetts: tete0@adventure-works.com
       Michigan: michael9@adventure-works.com
      Minnesota: jillian0@adventure-works.com
        Ontario: josé1@safe-mail.net
         Oregon: pamela0@yahoo.com
      Tennessee: tsvi0@adventure-works.com
           Utah: linda3@adventure-works.com
       Victoria: lynn0@adventure-works.com
     Washington: david8@adventure-works.com, stephen0@adventure-works.com, amy0@yahoo.com, syed0@yahoo.com, brian3@aol.com
<pre>

## Problem 6 (Bonus)
Define a function `managementCounts` which produces a Series of how many employees report to a manager. The Series is indexed by the `ManagerID`, the count should be performed on the `EmployeeID` because this is the only field that is guaranteed to be unique. The resulting Series should be order by the number of employees in **descending order**.

Arguments:
- employees (DataFrame)

Returns:
- Series of counts (int), indexed by `ManagerID`

In [164]:
employees_df.columns

Index(['ManagerID', 'TerritoryID', 'Title', 'FirstName', 'MiddleName',
       'LastName', 'Suffix', 'JobTitle', 'NationalIDNumber', 'BirthDate',
       'MaritalStatus', 'Gender', 'HireDate', 'SalariedFlag', 'VacationHours',
       'SickLeaveHours', 'PhoneNumber', 'PhoneNumberType', 'EmailAddress',
       'AddressLine1', 'AddressLine2', 'City', 'StateProvinceName',
       'PostalCode', 'CountryName'],
      dtype='object')

In [163]:
# step 1 pick out employeeids based on manager id
m_id = 250.0

employee_df.loc[employee_df['ManagerID'] == m_id ,['EmployeeID']]


KeyError: "None of [Index(['EmployeeID'], dtype='object')] are in the [columns]"

In [184]:
# def managementCounts(employees):
employees_df = employees_df.reset_index()
# employees_df = pd.read_excel('/home/data/AdventureWorks/Employees.xls')
employees_df.head()
employees_df['EmployeeID']

0      259
1      278
2      204
3       78
4      255
5       66
6      270
7       22
8      161
9      124
10      10
11     248
12     175
13     155
14     144
15     192
16     166
17     112
18     203
19     245
20     257
21     244
22      69
23     210
24     163
25     118
26     176
27     147
28     138
29      95
      ... 
261     70
262    219
263    148
264    140
265    290
266      5
267    252
268    266
269    233
270     83
271    251
272    241
273     25
274    108
275    239
276     27
277    186
278     38
279    154
280    114
281    123
282     29
283    101
284    221
285     18
286     46
287    202
288    289
289    288
290    999
Name: EmployeeID, Length: 291, dtype: int64

In [76]:
# step 2: group em_id by m_id
my_group = employees_df.groupby('ManagerID')['EmployeeID']

# loop thru the list and count unique size 
employees_df.groupby('ManagerID')['EmployeeID'].apply(lambda em_id_list: len(em_id_list.unique()))

# sorting 
employees_df.groupby('ManagerID')['EmployeeID'].apply(lambda em_id_list: len(em_id_list.unique())).sort_values(ascending=False)

ManagerID
26.0     178
25.0      30
250.0     17
274.0     10
263.0      9
249.0      9
16.0       8
1.0        8
227.0      6
235.0      5
287.0      3
273.0      3
234.0      3
285.0      1
Name: EmployeeID, dtype: int64